### ISSUE TITLE
[L4D] "Assert: **** OUT OF MEMORY! attempted allocation size: 0 ****"

### ISSUE THREAD

**Issue by TheRenegadist** [ValveSoftware/Source-1-Games#2646]
Getting an assert crash in Left 4 Dead which happens as soon as any map opens, this is usually fixed by re-joining once the map has fully started though if the server has 3/4 players it thinks you're still connected and thus you can't reconnect for a couple minutes. This doesn't happen all the time but I'd say it happens 8/10 map loads.

I am using a modified StudioMDL to bypass the 32 material limit for skinfamilies in a couple models which seems to be confusing the engine which may be leading to it running out of memory though this is just a theory. This also happens on occasion even without using models in-game with the 32 material limit bypass but it is much more rare.

 Very much hoping this is an easy fix on your (Valve's) end since you have access to the source code of the game and can probably expand the memory the game needs if that is the issue. Bare in mind my knowledge on the error is very limited and may be much more harder to fix.

Screencap of the error: https://media.discordapp.net/attachments/186995831986061313/483440452910055424/Desktop_8-26-2018_5-50-47_PM-367.png?width=450&height=253
Modified StudioMDL: https://dev.cra0kalo.com/?p=21

Many thanks!

**Comment by int19h** [comment]
Seeing this periodically as well, but I'm not using any overrides. The link above is dead, so I'll try to get a screenshot of the assert next time it happens

One thing of note is that the allocation size is 0. I wonder if this is happening because the allocator was updated such that it returns NULL for 0-size allocations (this is legal in standard C and C++), but the code that calls it incorrectly assumes that NULL indicates failure? E.g. the VC++ runtime was updated?

**Comment by TheRenegadist** [comment]
I realize this may be an engine limitation but I can load hundreds upon hundreds of HD textures, models, etc with Left 4 Dead 2 just fine but in L4DI'm barely scraping along with stuff that isn't half the size or quality.

It would at least be nice if this was also forwarded to L4D but I assume Valve is done touching L4D.

**Comment by int19h** [comment]
I don't think this is related to engine limitations. I'm seeing it in completely unmodded games, and on a machine that has gigabytes of RAM free.

**Comment by TheRenegadist** [comment]
Got another screencap since the old one expired:

https://i.imgur.com/1IReDie.png

**Comment by TheRenegadist** [comment]
Switched to an unmodded game and this is still happening as int19h said.

**Comment by TheRenegadist** [comment]
I'd like to stress again that this isn't only happening to me, a buddy of mine is trying to run Left 4 Dead default with nothing in the files touched and is still experiencing crashing, this isn't just an issue for people modding the game but both unmodded and modded copies of Left 4 Dead are suffering from this and no response has been issues by Valve.

His screencap: https://i.imgur.com/ts5jwny.jpg

Feel free to prove me wrong or call me on it but I know you guys don't care for Left 4 Dead anymore but if people buying the game are trying to run it default and it's crashing on them with the same error in the same code then shouldn't this be seriously looked into and hopefully fixed? I opened this on Auguest 26 and here we are almost a year later and nothing has been done, can you guys at least look into it?

**Comment by int19h** [comment]
After the most recent L4D update, this 1) happens more often - now also while trying to take screenshots, 2) and the allocation size is usually non-0 now.

And, of course, it's impossible to debug, because the binary is obfuscated, so nobody but Valve can fix it. I have a crash dump I could share though, if anybody is even reading these...

**Comment by TheRenegadist** [comment]
Not sure why I keep trying since it seems Valve either does not care to fix this but with the help of some other people in the community we've isolated the issue. 

It seems the crashing is happening when "Multicore Rendering" is enabled, disabling it does fix the crashing seen above but on custom campaigns and maybe even on official ones with more assets you get a new type of crash that appears as a window however you can still move around in-game and the game only closes once you select one of the three options. Screenshot below:

https://i.imgur.com/qBvbIzE.png

A reminder that this happens to both people with mods and custom campaigns and completely untouched directories, this NEEDS to be fixed because your customers are getting screwed over by this.

**Comment by v1993** [comment]
@TheRenegadist Wow, thank you so much! I've been unable to play Portal 2 with exactly same message and turning multicore rendering fixed the crash.

**Comment by benlavender** [comment]
Was happening to me on L4D unmodded on Windows 10. I've gone with what @TheRenegadist suggested and will update later.

**Comment by JayBoom** [comment]
Still happening 5/01/2023

**Comment by AshThe9thSurvivor** [comment]
> Sigue sucediendo 5/01/2023
The bug hasn't been fixed to this day, it's not even hard and it takes VALVE forever to fix something so simple.
The worst thing is that I offered to make a community update for L4D1, since I have experience doing it, since 2014 I am maintaining an unofficial update.

**Comment by AshThe9thSurvivor** [comment]
> Sigue sucediendo 5/01/2023

The bug hasn't been fixed to this day, it's not even hard and it takes VALVE forever to fix something so simple.
The worst thing is that I offered to make a community update for L4D1, since I have experience doing it, since 2014 I am maintaining an unofficial update.

**Comment by JayBoom** [comment]
> > Sigue sucediendo 5/01/2023
> 
> The bug hasn't been fixed to this day, it's not even hard and it takes VALVE forever to fix something so simple. The worst thing is that I offered to make a community update for L4D1, since I have experience doing it, since 2014 I am maintaining an unofficial update.

Have you tried posting on steam forums and contacting Valve employees through their contact website? I don't really know what else you can do though.

**Comment by AshThe9thSurvivor** [comment]
> 
> Have you tried posting on steam forums and contacting Valve employees through their contact website? I don't really know what else you can do though.
> 
> You can also try a steam support enquiry, won't hurt.

I have sent a message to VALVE's email, the one that appears inside the game, I made a post on Steam, in the area where the creators supposedly review, I even sent it to Steam support, I also post it here, they always tell me the same thing: "This is not the place where this type of post should be made, you should post it in the DISCUSSIONS area, there the creators check and get ideas for updates."
As I already mentioned, I have also published it there, up to 3 times, and it is a lie, they never review anything there.

You can see my latest post here: https://steamcommunity.com/app/500/discussions/0/3392923906934729525/

**Comment by JayBoom** [comment]
Thought turning off multi core rendering would fix it.. apparently not https://i.imgur.com/U4GIAuf.png

**Comment by AshThe9thSurvivor** [comment]
> Thought turning off multi core rendering would fix it.. apparently not https://i.imgur.com/U4GIAuf.png and also a new one I haven't seen: https://i.imgur.com/9McGmPv.png

Many things to correct, not a single update solving it.

**Comment by DyingJKJ** [comment]
still happenin on 23rd Jun 2024 XP

**Comment by Eugene01234** [comment]
> still happenin on 23rd Jun 2024 XP

yeah, i've got it too

**Comment by Eugene01234** [comment]
i think problem is in version of a game, 32 bit, and i think maximum ram this game using by this version is 3 GB

**Comment by jbiel** [comment]
This is actually a really easy fix and I'm disappointed that Valve hasn't done anything about it. I found a [relevant wiki page](https://www.pcgamingwiki.com/wiki/Left_4_Dead#Issues_fixed) and the fix outlined works for me.

**Comment by Eugene01234** [comment]
можно им написать чтобы они исправили

**Comment by Eugene01234** [comment]
here link to Patch => [https://ntcore.com/4gb-patch/](steam://openurl_external/https://steamcommunity.com/linkfilter/?u=https%3A%2F%2Fntcore.com%2F4gb-patch%2F)
